bindesc: Add initial support for binary descriptors

yonsch opened the pull request:
### Binary Descriptors
Binary Descriptors are strings and integers that describe a binary executable. With a bit of linker trickery, these descriptors are placed at a known location in the binary. This allows a host tool to read these descriptors, even if only a `bin` file is present. In the future it may allow different images on the same device to read each other's descriptors. This is useful mostly for interaction between an app and a bootloader - a bootloader reading the app's version, for example.

Binary Descriptors were inspired by `pico-sdk`'s [binary_info](https://github.com/raspberrypi/pico-sdk/tree/master/src/common/pico_binary_info).

### Working principles
Binary descriptors are implemented with a TLV (tag, length, value) header linked to a known offset in the binary image. This offset may vary between architectures, but generally the descriptors are linked as close to the beginning of the image as possible. In architectures where the image must begin with a vector table (such as ARM), the descriptors are linked right after the vector table. The reset vector points to the beginning of the text section, which is after the descriptors. In architectures where the image must begin with executable code (e.g. x86), a jump instruction is injected at the beginning of the image, in order to skip over the binary descriptors, which are right after the jump instruction.

A user can define any ID for any use case, but this will most likely be used for version strings and build timestamps. For these typical use cases, IDs are defined in `bindesc.h`. Also, default declarations for timestamp and version are provided and can be enabled with Kconfig (see `bindesc_build_time.c` as an example).

### Platform support
Binary descriptors have been proven to be pretty portable on various QEMU machines, but more testing needs to be done in this area.

| Platform | Builds | west bindesc works |  Runs  (renode/qemu)   | Runs (hardware) |
|----------|----------|-------|---|----------|
| ARM Cortex-M     | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:  | :heavy_check_mark:  |
| native_posix      | :heavy_check_mark: | :heavy_check_mark: | N/A | :heavy_check_mark: |
| RISC-V | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:  |  |
| ARC | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:  | |
| x86 | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:  | |

### C standard support
| Standard | Supported          |
|----------|--------------------|
| c99      | :heavy_check_mark: |
| c11      | :heavy_check_mark: |
| c17      | :heavy_check_mark: |
| gnu99    | :heavy_check_mark: |
| gnu11    | :heavy_check_mark: |
| gnu17    | :heavy_check_mark: |
| c++17      | :heavy_check_mark: |

### Toolchain support
| Toolchain | Supported          |
|----------|--------------------|
| zephyr-sdk (GCC)      | :heavy_check_mark: |

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

Laczen commented:
@yonsch, I like the idea.

Regarding the format of the binary info entries maybe this could be reworked to a tag, length, value format as this would allow whatever binary data to be stored.

I am also interested in any remarks regarding portability.

yonsch commented:
UPDATE: I've created another sample that reads its own descriptors and ran it on QEMU (I don't know of any way to run both a bootloader and an app on the same emulation). I could verify that this works for ARM, RISC-V (32 and 64 bit), x86 and ARC.
For x86 where execution starts at the beginning of the executable, a jump instruction is added that skips over the binary descriptors.

attie-argentum commented on include/zephyr/bin_descriptors.h:
```diff
@@ -0,0 +1,158 @@
+ * @brief A union for holding the value of the binary descriptor, which
+ * can be either an integer or a string.
+ */
+union bin_descriptor_value {
+	/** Value as an integer */
+	uint32_t integer;
+	/** Value as a pointer to a char */
+	const char *str;
+};
```
I'd also suggest reworking this to `type` / `len` / `data`... this way knowing the start address only will permit you "walk" through all the items.

If non-string / BLOB types (e.g: `uint32_t`, pointers, etc...) are really expected, then either special handling or alignment becomes important.

```suggestion
struct bin_descriptor {
	uint32_t type; /* e.g: BIN_DESC_TYPE_APP_VERSION_STRING */
	uint32_t len; /* length of data, below */
	uint8_t data[];
};
```

attie-argentum commented on include/zephyr/bin_descriptors.h:
```diff
@@ -0,0 +1,158 @@
+ * the executable pointed to does not support binary descriptors or because the address
+ * was invalid.
+ */
+int bin_descriptors_foreach(uint32_t *address, bin_descriptors_foreach_cb callback,
+		void *user_data);
```
I'd probably also appreciate a "_find me this descriptor_" type interface...

```
int bin_descriptors_get(uint32_t *address, uint32_t type, uint32_t index, const struct bin_descriptor **desc);
```

attie-argentum commented on subsys/bin_descriptors/bin_descriptors.ld:
```diff
@@ -0,0 +1,7 @@
+
+LONG(BIN_DESCRIPTOR_MAGIC);
+LONG((_bin_desc_end - _bin_desc_start) / 8);
+_bin_desc_start = .;
```
See above regarding alignment.

attie-argentum reviewed (dismissed):
This looks good, thanks for your effort!

> The data can be either an integer or a pointer to a string somewhere in RODATA.

I'd prefer for this to be a contiguous and self-contained block. Once you've found the start, you can walk through the rest of it without pulling data from elsewhere or seeking into the binary. (i.e: no pointers)

Tooling to extract and display this information on a PC would also be incredibly useful.

---

I'd also be very keen for this to gain an equivalent to GNU's `--build-id=sha1` (see #51532, [ref](https://interrupt.memfault.com/blog/gnu-build-id-for-firmware)), which is a hash over all "_interesting_" parts of the binary - giving confidence when matching running code to build artifacts, for debugging, etc...

That's non-trivial to get right, so being able to expand into that is fine.

yonsch commented:
@attie-argentum Thanks for your comments.

- I've changed the structure to id/len/data. It introduces some more alignment issues, but I also think it's better (unless this would cause a problem down the road where the header would get too large).

- A python script for displaying the descriptors was also added. It's very clumsy but I'll make it better when we're settled on the header format. I hope to have this as a west extension, and in the future to connect it with a runner so that it would be possible to tell what's currently flashed on a device.
- A function for searching a specific descriptor was added.
- Regarding the hash - is that a hash over the entire image? I don't think it's possible to calculate this at link time, so perhaps we can fill it with zeros, and then have a post build script replace it with the correct hash.

ddavidebor commented:
Love this idea! 
If you end up merging it, please consider adding documentation that explains how to sneak in these Binary Descriptors from environment variables.

yonsch commented:
@ddavidebor thanks, I will add a more thorough documentation once the general idea is approved. I would say that environment variables should make their way into the image through cmake: `zephyr_compile_definitions($ENV{...})`.
Were you thinking of any particular variable? I might add it to the common IDs...

ddavidebor commented:
@yonsch 

Yeah, some useful variables could be

- toolchain version
- some sort of computer ID
- GitHub commit hash
- CI build number

JonathanVanWin commented:
Interesting concept @yonsch .
We also use something similar to this at my place, exactly for storing the compiled version of the binary and other stuff.
How much place do we allocate for this IDs?
What if we need more place, can we change how much memory is allocated for this?

attie-argentum commented:
@yonsch 

> Regarding the hash - is that a hash over the entire image?

Yes, the "functional" pieces.

> I don't think it's possible to calculate this at link time, so perhaps we can fill it with zeros, and then have a post build script replace it with the correct hash.

I'm not sure how exactly `ld` calculated the "_Build ID_", but it's an approach that would be worth reproducing. Zeroes with binary patch sounds plausible, but a little risky.


@ddavidebor 

> GitHub commit hash

It would also be handy to have the "application" repository's commit ID too... and perhaps clean/dirty flags for the sources.

---

I include the following JSON BLOB (or similar) in my applications - put together using a script, and with tools to extract the info from a binary/ihex. Some of the same information is also available on the console output, though not currently parsed from the JSON at runtime.

I tend to use my application's repository as the "authority", with it's own `west.yml`, etc... so all sources can be found if built from a clean tree. I think the commit ID and clean/dirty indication of supporting projects is also important.

The release name/date comes from `git describe --tag --exact-match --match 'v*.*' --dirty='+'` from the application (not Zephyr).

```json
{
	"product": $product_name,
	"build": {
		"date":        $build_date,
		"gcc":         $build_gcc_version,
		"zephyr-sdk":  $build_zephyr_sdk_version
	},
	"release": {
		"name":        $release_name,
		"date":        $release_date,
	},
	"firmware": {
		"version":     $firmware_version,
		"config":      $firmware_config
	},
	"source": {
		"application": $commit_application,
		"zephyr":      $commit_zephyr,
		"hal/cmsis":   $commit_hal_cmsis,
		"hal/atmel":   $commit_hal_atmel,
		"fs/fatfs":    $commit_fs_fatfs
	}
}
```

yonsch commented:
@attie-argentum 

> Zeroes with binary patch sounds plausible, but a little risky, especially if you're acting on a single file (rather than different filenames for pre/post patch).

Maybe this can be the hash of one of the intermediate ELFs or some of their content. Then we can generate the hash as a *.c or *.o file and link it into the final binary.

Thanks for the other ideas, I'll add IDs for them

yonsch commented:
Pinging some of the participants of #51532 , if the general idea is approved I'll get this finalized for a review:
@JordanYates @nordicjm @ceolin @tejlmand @stephanosio

nordicjm commented:
The underlying storage has quite a few parts in common with a data retention API I am developing, I wonder if they could be merged. I wonder if putting the position into dts would be a better fit also, the position might want to be configurable by someone, e.g. not to be at the beginning of an application but somewhere else, or they might want multiple sections at different offsets.

I like the idea!

Laczen commented:
@yonsch, as you know I like the idea. It could be used by the build system to find out image properties when combining images.

However as I was thinking into using the feature I stumbled on a problem. To allow getting for example the version of the bootloader the properties are placed behind the vector table (on arm). When using this with encrypted images this means that the properties would become encrypted and are no longer readable by external programs that don't have an idea of the encryption key. There is a solution to this by placing the properties before the vector table (so that they can be left unencrypted), but this means that the properties can no longer be used to find out the version of the bootloader (as the bootloader cannot add image data before the vector table). But then again maybe it is not bad to disallow reading from the bootloader al together as this might provide means to discover bootloader secrets.

What is your opinion on placing the binary descriptors before the vector table ?

attie-argentum commented:
> What is your opinion on placing the binary descriptors before the vector table ?

I think you'll find that the subject of "_before/after vector table_" is not directly compatible with other architectures... Placing this information as a block at "_the front_" of the image (generally speaking) would permit a small amount to be decrypted before locating this information in full.

If you need to inspect details of an encrypted image without the keys, then I'd suggest this becomes a packaging issue - i.e: a disposable header followed by the encrypted payload / firmware image would be a preferable approach... thoughts?

Laczen commented:
> > What is your opinion on placing the binary descriptors before the vector table ?
> 
> If you need to inspect details of an encrypted image without the keys, then I'd suggest this becomes a packaging issue - i.e: a disposable header followed by the encrypted payload / firmware image would be a preferable approach... thoughts?

@attie-argentum I guess you are right, it would be possible to extract the (limited amount of) needed data from a image and put it in a (disposable) header. It would duplicate the data but when it is extracted from the image it is certain not to be different.

yonsch commented:
@nordicjm
> I wonder if putting the position into dts would be a better fit also, the position might want to be configurable by someone, e.g. not to be at the beginning of an application but somewhere else, or they might want multiple sections at different offsets.

The problem I see with this is that any app trying to read the descriptors would also need the dts to know where to look for them. If the descriptors are always at the same location you always know where to find them.

@Laczen  @attie-argentum 
I really haven't thought about encryption, and my knowledge in that area is pretty limited. I suppose that whatever tool is used to encrypt the image skips the first X bytes of the image, as they might include a header and the vector table. In that case, we only have to tell that tool to also skip the descriptors. But again, I've never done this before so I have no idea how hard this might be.

nordicjm commented:
> The problem I see with this is that any app trying to read the descriptors would also need the dts to know where to look for them. If the descriptors are always at the same location you always know where to find them.

That's no different than now. You can't build mcuboot and an mcuboot image without that otherwise how does mcuboot know where the image is, what the slot size is, etc.?

carlescufi commented:
Architecture WG:

- @henrikbrixandersen mentions that this functionality is implemented in their downstream
- @Laczen mentions that it might be a security risk to be able to read the bootloader. So one of the biggest use cases, being able to get which mcuboot image version, is by default insecure because it requires read access to the bootloader

carlescufi commented:
> Architecture WG:
> 
> * @henrikbrixandersen mentions that this functionality is implemented in their downstream, with the objective of being able to find out the mcuboot version from the app
> * @Laczen mentions that it might be a security risk to be able to read the bootloader. So one of the biggest use cases, being able to get which mcuboot image version, is by default insecure because it requires read access to the bootloader. That said, @Laczen mentions that he thinks it's a good proposal

@yonsch given the feedback above, I would recommend you take this out of draft.

nordicjm commented:
@yonsch I've created this PR for adding an application version Kconfig option which will work with imgtool signing and might be of use for this too, https://github.com/zephyrproject-rtos/zephyr/pull/55814 can you have a look and see if you think if it would be a good fit?

yonsch committed:
bindesc: Add initial support for binary descriptor definition

Binary descriptors are data objects stored at a known location
of a binary image. They can be read by an external tool or image,
and are used mostly for build information: version, build time,
host information, etc.
This commit adds initial support for defining such descriptors.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch committed:
west: commands: Add bindesc command

Added the bindesc command to west, for working with binary
descriptors. Currently it supports dump, list and search
subcommands, for bin, hex, elf and uf2 file types.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch committed:
samples: bindesc: Add hello_bindesc sample

Add the hello_bindesc sample which shows the basic usage of
binary descriptors.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch committed:
doc: bindesc: Add documentation for binary descriptors

Add documentation for binary descriptors under "OS Services"

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch commented:
# Update
I've marked this pull request as ready for review.

For now, I've dropped support for reading the descriptors from another image, as it raised some security concerns. I will open a separate PR in the future adding this feature, after we'll have a discussion on how to implement it securely. In the meantime, the given proposal providing a means to define descriptors and show them with the `west` tool is still useful on its own.

mbolivar-nordic commented on subsys/bindesc/Kconfig.version:
```diff
@@ -0,0 +1,38 @@
+config BINDESC_KERNEL_VERSION_PATCHLEVEL
+	bool "Kernel version patchlevel"
+	help
+	  The patchlevel version number, such as 3 in 3.4.0
```
0, not 3

mbolivar-nordic commented on subsys/bindesc/Kconfig.version:
```diff
@@ -0,0 +1,38 @@
+config BINDESC_KERNEL_VERSION_STRING
+	bool "Kernel version string"
+	help
+	  The kernel version string, such as "3.4.0"
```
consistency nitpick: this "3.4.0" is quoted but none of the other ones below are quoted

mbolivar-nordic commented on subsys/bindesc/Kconfig.version:
```diff
@@ -0,0 +1,38 @@
+	help
+	  The kernel version as an integer computed as
+	  (major << 16 | minor << 8 | patchlevel). For example,
+	  3.4.0 would be represented as 0x30400
```
Mentioning "binary coded decimal" here might be nice.

mbolivar-nordic commented on scripts/west_commands/bindesc.py:
```diff
@@ -0,0 +1,284 @@
+from west.commands import WestCommand
+from west import log
+
+from elftools.elf.elffile import ELFFile
+from intelhex import IntelHex
```
Please follow the MISSING_REQUIREMENTS pattern in runners/jlink.py to gracefully handle missing non-standard-library, non-west imports.

mbolivar-nordic commented on scripts/west_commands/bindesc.py:
```diff
@@ -0,0 +1,284 @@
+    MAGIC = 0x2c31b9b9
+    DESCRIPTORS_END = 0xffffffff
+
+    TYPE_UINT = 0
+    TYPE_STR = 1
```
These and the corresponding definitions in the C header should be commented to say "if you change this, you need to keep $THE_OTHER_ONE in sync". Similar comment for any other shared constants.

mbolivar-nordic commented on scripts/west_commands/bindesc.py:
```diff
@@ -0,0 +1,284 @@
+
+        if self.file_type == 'uf2':
+            with open(file_name, 'rb') as uf2_file:
+                return convert_from_uf2(uf2_file.read())
```
Please rework this to avoid use of `assert` (use RuntimeError instead) and `print` (use `log`)

mbolivar-nordic commented on doc/services/binary_descriptors/index.rst:
```diff
@@ -0,0 +1,127 @@
+
+For more information read the ``help`` sections of these Kconfig symbols.
+
+west bindesc tool
+=================
+``west`` is able to parse and display binary descriptors from a given executable image.
+It currently supports ``.bin``, ``.hex``, ``.elf`` and ``.uf2`` files. The supported
+subcommands are:
+
+ 1. ``search`` - Prints the value of a specific descriptor found in an image by name
+ 2. ``custom_search`` - Prints the value of a specific descriptor found in an image by type and ID
+ 3. ``dump`` - Prints the values of all descriptors found in an image
+ 4. ``list`` - Prints all known standard descriptor tags
+
+For more information refer to ``west bindesc --help``.
```
Please move this under doc/develop/west instead, and add cross-references to that page from here, and to this page from there.

mbolivar-nordic commented on doc/services/binary_descriptors/index.rst:
```diff
@@ -0,0 +1,127 @@
+.. _binary_descriptors:
+
+Binary Descriptors
```
What is the current API stability level?

yonsch replied:
Hmm, what do you think? I hadn't had any chance to use it in a real application yet. I don't think it'll change but who knows...

mbolivar-nordic reviewed (changes requested):
A lot of work here; nice job. I noticed a few things. I am +1 on the idea and adding a west extension for it.

mbolivar-nordic commented on doc/develop/west/zephyr-cmds.rst:
```diff
@@ -185,3 +185,28 @@ Twister can then be invoked via west as follows::
+Working with binary descriptors: ``west bindesc``
+*************************************************
+
+The ``bindesc`` command allows users to read :ref:`binary descriptors<_binary_descriptors>`
```
This won't work; should be `<binary_descriptors>` (no `_`)

mbolivar-nordic commented on doc/services/binary_descriptors/index.rst:
```diff
@@ -0,0 +1,120 @@
+=================
+``west`` is able to parse and display binary descriptors from a given executable image.
+
+For more information refer to ``west bindesc --help`` or the :ref:`documentation<_west-bindesc>`.
```
`<west-bindesc>`

yonsch commented:
> > the actual names of the variables used to access the built-in descriptors (kernel_version_string, kernel_version_major, etc.) do not appear to be documented anywhere
> 
> I think it would be enough to document in the .rst that the variable name is the lowercased kconfig option name with `CONFIG_` stripped off, along with a couple of examples

@mbolivar-nordic I've added the variables to doxygen

mbolivar-nordic reviewed (dismissed):
Docs look good enough to me, and I think the rest of this feature is well baked.

I think it would be an improvement to the documentation to:

- put the variables in a separate doxygen group or groups
- pull that group into the .rst with some introductory text about the Kconfig options that control the visibility of those symbols. Without that, I expect some people to get confused about why they get linker errors since they won't know they have to enable the options unless they study the source code.

The approach I'm suggesting is similar to what we do in the devicetree API. See for example:

https://github.com/zephyrproject-rtos/zephyr/blob/600e7b7f21cfd179609372bb4f5e3bc5d33d19c2/doc/build/dts/api/api.rst?plain=1#L49

But that's not a blocker; this is definitely already much better documentation than zephyr's usual quality.

Nice work!

attie-argentum reviewed (commented):
Looking great, well done!

A few comments / probes from me... nothing worth blocking over, but perhaps things worth discussing to "_get it right_" first time(? - ha)

yonsch commented on include/zephyr/bindesc.h:
```diff
@@ -0,0 +1,369 @@
+#define BINDESC_TAG_BUILD_TIME_MONTH BINDESC_TAG(UINT, 0x3001)
+
+/** The day of the month the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_DAY BINDESC_TAG(UINT, 0x3002)
+
+/** The hour of the day the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_HOUR BINDESC_TAG(UINT, 0x3003)
+
+/** The minute the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_MINUTE BINDESC_TAG(UINT, 0x3004)
+
+/** The second the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_SECOND BINDESC_TAG(UINT, 0x3005)
+
+/** The UNIX time (seconds since midnight of 1970/01/01) the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_UNIX BINDESC_TAG(UINT, 0x3006)
+
+/** The date and time of compilation such as "2023/02/05 00:07:04" */
+#define BINDESC_TAG_BUILD_DATE_TIME_STRING BINDESC_TAG(STR, 0x3007)
+
+/** The date of compilation such as "2023/02/05" */
+#define BINDESC_TAG_BUILD_DATE_STRING BINDESC_TAG(STR, 0x3008)
+
+/** The time of compilation such as "00:07:04" */
+#define BINDESC_TAG_BUILD_TIME_STRING BINDESC_TAG(STR, 0x3009)
+
+/** The name of the host that compiled the image */
+#define BINDESC_TAG_HOST_NAME BINDESC_TAG(STR, 0x4000)
+
+/** The C compiler name */
+#define BINDESC_TAG_C_COMPILER_NAME BINDESC_TAG(STR, 0x4001)
+
+/** The C compiler version */
+#define BINDESC_TAG_C_COMPILER_VERSION BINDESC_TAG(STR, 0x4002)
+
+/** The C++ compiler name */
+#define BINDESC_TAG_CXX_COMPILER_NAME BINDESC_TAG(STR, 0x4003)
+
+/** The C++ compiler version */
+#define BINDESC_TAG_CXX_COMPILER_VERSION BINDESC_TAG(STR, 0x4004)
```
I started it from 0x1000 so that if a careless user accidentally uses 0/1/2/3... it won't cause any collision. The thing is that 32 bits is a huge field compared to what is stored in it. I really am thinking of shrinking the length (another overkill) and tag to 16 bits each. The only reason I haven't done this yet is that I fear this would cause issues with alignment or packing down the road. But I think with the proper flags and build assertions we can make it work and even save some space.

attie replied:
Fair enough... explicit reservations are probably still a good idea

attie-argentum replied:
Fair enough... explicit reservations are probably still a good idea?

yonsch commented on include/zephyr/bindesc.h:
```diff
@@ -0,0 +1,369 @@
+		.tag = _tag,			\
+		.len = sizeof(value),		\
+		.data = value,			\
+	}
```
Great idea, thanks. As you mentioned the BINDESC_TAGS_* defs should now only be IDs, so they were renamed to BINDESC_ID_*

yonsch commented:
Update: the tag and length were reduced to 16 bits each. Several platforms and C standards were tested (see tables in the description)

yonsch committed:
tests: bindesc: Added definition tests

Added tests for the bindesc subsystem, testing definition of
binary descriptors on several qemu architectures, and using
several C standards (c99, c11, etc.).

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

nordicjm commented on subsys/bindesc/Kconfig.version:
```diff
@@ -0,0 +1,38 @@
+
+if BINDESC_DEFINE_VERSION
+
+config BINDESC_KERNEL_VERSION_STRING
```
This is already generated from the zephyr VERSION file, it should be used instead instead of duplicating here. For an *application* version there is an RFC for adding it https://github.com/zephyrproject-rtos/zephyr/issues/56605 with 2 possible implementations, likewise this should use that rather than duplicating it

yonsch replied:
This is not the actual version, this is a boolean that selects whether you want it as a descriptor or not

carlocaione reviewed (changes requested):
How does this co-exist with `CONFIG_AARCH64_IMAGE_HEADER` and `CONFIG_ARM_ZIMAGE_HEADER`?

carlocaione reviewed (dismissed):
How does this co-exist with `CONFIG_AARCH64_IMAGE_HEADER` and `CONFIG_ARM_ZIMAGE_HEADER`?

yonsch commented:
> How does this co-exist with `CONFIG_AARCH64_IMAGE_HEADER` and `CONFIG_ARM_ZIMAGE_HEADER`?

Currently this feature relies on ROM_START as an argument to zephyr_linker_sources, so it's limited to architectures supporting it. ARM64 is not supported yet, so it wouldn't work anyway.
I'm not familiar with ARM64 but from a glance it looks like that section could go after the image header and vector table, so it shouldn't be a problem.

nashif commented on subsys/bindesc/Kconfig:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2023 Yonatan Schachter
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig BINDESC
```
Is this really a subsystem?

yonsch replied:
It's really more of a build system feature, so I though the most proper place would be with the subsystems. But if you think a different location would be more suitable, I'll move it.
Also note that in the future we might add code for reading those descriptors from the device itself, if that makes a difference.

carlocaione commented:
> ARM64 is not supported yet, so it wouldn't work anyway.

If this is not going to be supported on all architectures then we need at least an `ARCH_HAS_` Kconfig symbol of sorts to filter architectures out.

yonsch committed:
arch: Add ARCH_SUPPORTS_ROM_START kconfig symbol

Add an ARCH_SUPPORTS_ROM_START kconfig symbol to mark architectures
that support ROM_START as an argument to zephyr_linker_sources.
This was added so that features relying on this feature could
depend on this kconfig symbol.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

nordicjm commented on subsys/bindesc/Kconfig.version:
```diff
@@ -0,0 +1,38 @@
+	help
+	  The kernel version as binary coded decimal, computed as
+	  (major << 16 | minor << 8 | patchlevel). For example,
+	  3.4.0 would be represented as 0x30400
```
Can add application version too

yonsch replied:
This is currently problematic, see https://github.com/zephyrproject-rtos/zephyr/pull/56141#issuecomment-1579143844

nordicjm commented on subsys/bindesc/Kconfig:
```diff
@@ -0,0 +1,25 @@
+	help
+	  Enable the app to define its own binary descriptors
+
+if BINDESC_DEFINE
```
Shouldn't this not be here, as version, build time, etc. would not be defining own descriptors, they are in-built ones?

yonsch replied:
This symbol is to differentiate between defining descriptors and reading them (in the future). For example, the bootloader would need to read the descriptors of the app without defining its own, and the app would define descriptors without needing to read any.

nordicjm commented on scripts/west_commands/bindesc.py:
```diff
@@ -0,0 +1,316 @@
+            index += length
+            index = self.align(index, 4)
+            current_tag = self.bytes_to_short(image[index:index+2])
+            # print("tag", current_tag)
```
Some comments left in the code

nordicjm reviewed (dismissed):
Looks good, not tried it but looked at the code, some minor nits.

nordicjm commented on include/zephyr/bindesc.h:
```diff
@@ -0,0 +1,402 @@
+#define BINDESC_ID_APP_VERSION_MINOR 0x802
+
+/** The app version revision such as 3 */
+#define BINDESC_ID_APP_VERSION_REVISION 0x803
```
Shouldn't this be 2 fields, patchlevel and number?

nordicjm commented on scripts/west_commands/bindesc.py:
```diff
@@ -0,0 +1,314 @@
+            self.bindesc_gen_tag(self.TYPE_STR, 0x800): 'APP_VERSION_STRING',
+            self.bindesc_gen_tag(self.TYPE_UINT, 0x801): 'APP_VERSION_MAJOR',
+            self.bindesc_gen_tag(self.TYPE_UINT, 0x802): 'APP_VERSION_MINOR',
+            self.bindesc_gen_tag(self.TYPE_UINT, 0x803): 'APP_VERSION_REVISION',
```
And here, I don't see this field, I see BINDESC_ID_APP_VERSION_PATCHLEVEL and BINDESC_ID_APP_VERSION_NUMBER

yonsch commented:
@nordicjm Thanks, I've added definitions for the app version number and also renamed the app version "revision" to "patchlevel", to comply with the naming used in `example-application`

carlescufi commented on subsys/bindesc/bindesc_skip.S:
```diff
@@ -0,0 +1,9 @@
+/*
```
This file should probably be named `bindesc_skip_x86.S` or, even better, x86/bindesc_skip.S`

carlescufi commented on include/zephyr/bindesc.h:
```diff
@@ -0,0 +1,405 @@
+ * @param id Unique ID of the descriptor
+ * @param value A string value for the descriptor
+ */
+#define DEFINE_BINDESC_STR(name, id, value)	\
```
Not a fan of the `DEFINE_` prefix. Not blocking, but could we drop it?

yonsch replied:
Originally this was to be consistent with the rest of Zephyr's API, but now I realize that the rest of the API uses it as a postfix (e.g. K_THREAD_DEFINE, K_TIMER_DEFINE, K_QUEUE_DEFINE etc.). So I can either remove it or change is to a postfix, whatever you decide.

de-nordic replied:
IMHO the <SUBSYS>_<OP> naming convention should be used. It is also more convenient when using auto completion because you type several letters of subsys and then select among provided defines only for the subsys.

yonsch replied:
> IMHO the _ naming convention should be used. It is also more convenient when using auto completion because you type several letters of subsys and then select among provided defines only for the subsys.

@de-nordic What's the _ naming convention? Do you mean that everything should start with `BINDESC_`?

carlescufi replied:
@yonsch let's switch to `_DEFINE` as a postfix.

carlescufi replied:
This is the last blocking bit from my side.

de-nordic replied:
> > IMHO the _ naming convention should be used. It is also more convenient when using auto completion because you type several letters of subsys and then select among provided defines only for the subsys.
> 
> @de-nordic What's the _ naming convention? Do you mean that everything should start with `BINDESC_`?

Yes, I think so.

carlescufi commented on include/zephyr/bindesc.h:
```diff
@@ -0,0 +1,405 @@
+
+/*
+ * Corresponds to the definitions in scripts/west_commands/bindesc.py.
+ * Do not change without syncing the definitions in both files!
```
It'd be great if the Python script parsed this header instead.

yonsch replied:
My concern here was about how the script would find that file. Looking for `ZEPHYR_BASE` might not work because the user may have another instance of Zephyr of an older version, and looking for it relative to the script may also not work if the script is running from a copied version (e.g. west is running from a venv)

carlescufi replied:
I don't think that running from a venv changes the location of the original script, only those of the pip modules that you are installing in the venv. But let's leave this as an improvement.

carlescufi commented on include/zephyr/bindesc.h:
```diff
@@ -0,0 +1,405 @@
+ * think it is. If these assertions fail for your toolchain/platform,
+ * please open a bug report.
+ */
+BUILD_ASSERT(offsetof(struct bindesc_entry, tag) == 0, "Incorrect memory layout");
```
These lines could be replaced with:
`BUILD_ASSERT(sizeof(struct bindesc_entry) == 4))`
no?

yonsch replied:
The entry may contain a string which is of arbitrary size

carlescufi replied:
not `struct bindesc_entry`, that only contains an empty array.

carlescufi replied:
Even if I'm right I don't mind if you leave this as-is, it's not blocking.

carlescufi commented on include/zephyr/bindesc.h:
```diff
@@ -0,0 +1,405 @@
+#define BINDESC_NAME(name) bindesc_entry_##name
+
+/*
+ * Utility macro to convert a uint32_t to a uint8_t array
```
I think you have all you need here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/byteorder.h

No need to add more macros?

carlescufi reviewed (commented):
Not blocking, just a few comments from me.

yonsch committed:
byteorder: Add macros for converting ints to arrays

Add three macros: sys_uint{16,32,64}_to_array, to convert
integers to byte arrays in a byte order aware manner.
For example, sys_uint16_to_array(0x0123) evaluates to:
{0x01, 0x23} for big endian machines, and {0x23, 0x01} for
little endian machines.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch committed:
tests: byteorder: Add tests for sys_uint*_to_array macros

Add tests for sys_uint*_to_array macros to the byteorder suite.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

dcpleung commented on scripts/west_commands/bindesc.py:
```diff
@@ -0,0 +1,315 @@
+
+        log.die('Unknown file type')
+
+    def parse_descriptors(self, image):
```
This probably needs to be more resilient to arbitrary hex block. The MAGIC sequence can appear in normal code and data outside of the binary descriptor block, especially in a bin file where you search through the whole binary. Or that a file with this MAGIC but not built with binary descriptor, etc... etc... etc. Probably a good idea to create a future enhancement ticket in the same time so we will remember to amend this.

yonsch replied:
Right, though the odds of a collision are pretty low as the descriptors are always very close to the beginning of the file (unless it's a native posix elf or a hex with multiple images).
I don't think this is much of a problem, because the odds are low and the script is not used for production. For the "real" use of this feature, which is reading the descriptors between partitions (will hopefully be added in a later PR), the offset will be known, and there will be no need to search for the magic.
If you still find it problematic, I can think of two solutions. The first is to add an offset argument to the script, which will make it much harder to use. The second is to make the magic much longer to make the odds of a collision negligible. In that case the change should be made now, to not break compatibility.

dcpleung replied:
A 64-bit magic can encode `BINDESC`. Makes it easier to look at the binary using hex editor. :)

Though, it is up to you. I am good either way.

yonsch replied:
I made the magic 64 bits long, though I kept it a random number (`BINDESC.` may turn up in a native posix ELF, in the file name or debug symbols or whatever)

nordicjm commented on samples/subsys/bindesc/bindesc.rst:
```diff
@@ -0,0 +1,10 @@
+.. _bindesc_samples:
```
needs a sample.yaml file - not blocking this PR as it is now but please ensure this compiles on a board you have locally, then submit a PR adding a sample.yaml file after this is merged

de-nordic commented on doc/services/binary_descriptors/index.rst:
```diff
@@ -0,0 +1,125 @@
+so if the image belongs to a big endian SoC, the appropriate flag should be given to the
+tool.
+
+The binary descriptor header starts with the magic number ``0xb9863e5a7ea46046``. It's followed
```
How the magic has been selected?

carlocaione replied:
I get a different value though (/jk 😜 )

de-nordic replied:
> I get a different value though (/jk 😜 )

I guess we are not doing this ;)
```
#define BINDESC_MAGIC hex(random.randint(0,0xffffffffffffffff)
```

> `hex(random.randint(0,0xffffffffffffffff)`

A little bit disappointed: I thought that you have figured out some magic opcode that is invalid for every architecture and will never be generated by compiler and was wondering what was your approach to get that...

de-nordic commented on include/zephyr/bindesc.h:
```diff
@@ -0,0 +1,398 @@
+ *
+ * @details
+ * Get the size of a binary descriptor. This is particularly useful for
+ * bytes binary descriptors where there's no null terminator.
```
Why do you need null terminator in string, you know the length already because you have encoded it anyway.

de-nordic replied:
> Why do you need null terminator in string, you know the length already because you have encoded it anyway.

OK I see now, to access from the code.

de-nordic commented on include/zephyr/bindesc.h:
```diff
@@ -0,0 +1,398 @@
+extern const struct bindesc_entry BINDESC_NAME(app_version_number);
+#endif /* IS_ENABLED(CONFIG_BINDESC_APP_VERSION_NUMBER) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_YEAR)
+extern const struct bindesc_entry BINDESC_NAME(build_time_year);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_YEAR) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_MONTH)
+extern const struct bindesc_entry BINDESC_NAME(build_time_month);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_MONTH) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_DAY)
+extern const struct bindesc_entry BINDESC_NAME(build_time_day);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_DAY) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_HOUR)
+extern const struct bindesc_entry BINDESC_NAME(build_time_hour);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_HOUR) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_MINUTE)
+extern const struct bindesc_entry BINDESC_NAME(build_time_minute);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_MINUTE) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_SECOND)
+extern const struct bindesc_entry BINDESC_NAME(build_time_second);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_SECOND) */
+
+#if IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_UNIX)
+extern const struct bindesc_entry BINDESC_NAME(build_time_unix);
+#endif /* IS_ENABLED(CONFIG_BINDESC_BUILD_TIME_UNIX) */
```
IMHO this is overkill. It should be enough to just have epoch string, which will be far more efficient than doing these separately.

yonsch replied:
I added them in case someone will ever find it useful. I agree that the individual year/month/day etc. is a very inefficient way to store the build time, but you don't pay for it if you don't use it so I see no harm in having them.

de-nordic commented on include/zephyr/bindesc.h:
```diff
@@ -0,0 +1,398 @@
+
+/** The kernel version number such as 0x30400 */
+#define BINDESC_ID_KERNEL_VERSION_NUMBER 0x904
+
```
Probably could allocate one for Zephyr git sha or app git sha.

de-nordic commented:
@yonsch Can you check whether it works correctly when you have an app that has been padded for use with bootloader, for example when you build it with the CONFIG_BOOTLOADER_MCUBOOT=y, and whether this still boots from bootloader?

de-nordic commented:
> > @yonsch Can you check whether it works correctly when you have an app that has been padded for use with bootloader, for example when you build it with the CONFIG_BOOTLOADER_MCUBOOT=y, and whether this still boots from bootloader?
> 
> @de-nordic
> 
> ```
> *** Booting Zephyr OS build zephyr-v3.2.0-14020-gc35d04a4b3c4 ***
> I: Starting bootloader
> I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Boot source: none
> I: Image index: 0, Swap type: none
> I: Bootloader chainload address offset: 0xc000
> I: Jumping to the first image slot
> *** Booting Zephyr OS build zephyr-v3.2.0-14020-gc35d04a4b3c4 ***
> Zephyr version: 3.4.99
> App version: 1.0.0
> Build time: 2023/09/26 14:14:18
> Compiler: GNU 12.2.0
> my_string: Hello world!
> my_int: 5
> my_bytes: 01 02 03 04
> ```

Thanks!

tejlmand commented on subsys/bindesc/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+      CONFIG_BINDESC_BUILD_TIME_STRING_FORMAT)
+
+  if(CONFIG_BINDESC_BUILD_TIME_ALWAYS_REBUILD)
+    add_custom_target(
+      bindesc_time_force_rebuild
+      COMMAND ${CMAKE_COMMAND} -U BUILD_TIME_YEAR ${CMAKE_BINARY_DIR}
+    )
```
why are you trying to remove a variable from the CMake cache which never seems to be added to the cache ?

Can you provide some details on your intentions.

yonsch replied:
@tejlmand The intention was to make cmake always rebuild `bindesc_build_time.c` so that the timestamp will be accurate. This is the solution I found somewhere, and I admit that now looking at it, `BUILD_TIME_YEAR` was not the variable name that was supposed to be there. I'm not sure if it got there by mistake in a refactor or that originally I intended for this cached value to be created for every time descriptor. Nevertheless - this solution is working. I changed `BUILD_TIME_YEAR` to `BUILD_TIME_DUMMY` (which is not set anywhere), and sure enough, when `CONFIG_BINDESC_BUILD_TIME_ALWAYS_REBUILD` is enabled it always rebuilds the file, and when it's disabled it does not.

carlescufi commented:
@attie-argentum please revisit

carlescufi commented:
@yonsch some twister build failures appeared in your last push

tejlmand commented on subsys/bindesc/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+    # done to ensure that the timestamp is always up to date.
+    add_custom_target(
+      bindesc_time_force_rebuild
+      COMMAND ${CMAKE_COMMAND} -U BUILD_TIME_DUMMY ${CMAKE_BINARY_DIR}
```
follow up to comment here: https://github.com/zephyrproject-rtos/zephyr/pull/54464#discussion_r1337379208

if it is purely a matter of re-invoking CMake, then there is no need for the `-U`, and this proposal should be sufficient:
```suggestion
      COMMAND ${CMAKE_COMMAND} ${CMAKE_BINARY_DIR}
```

that said, then i'm generally not fond of build system implementation where the following cannot be achieved when there are no changes to the system:
```
$ ninja
ninja: no work to do.
```
but in this case where the behavior is guarded by `CONFIG_BINDESC_BUILD_TIME_ALWAYS_REBUILD` then it is accepted.

nordicjm commented on scripts/west_commands/bindesc.py:
```diff
@@ -0,0 +1,316 @@
+            self.bindesc_gen_tag(self.TYPE_UINT, 0x801): 'APP_VERSION_MAJOR',
+            self.bindesc_gen_tag(self.TYPE_UINT, 0x802): 'APP_VERSION_MINOR',
+            self.bindesc_gen_tag(self.TYPE_UINT, 0x803): 'APP_VERSION_PATCHLEVEL',
+            self.bindesc_gen_tag(self.TYPE_UINT, 0x803): 'APP_VERSION_NUMBER',
```
*0x804

de-nordic replied:
Heh, good catch!

yonsch replied:
Very nice indeed

nashif commented on samples/subsys/bindesc/hello_bindesc/README.rst:
```diff
@@ -0,0 +1,26 @@
+
+.. code-block:: bash
+
+   west bindesc dump build/zephyr/zephyr.bin
```
would be nice to have more details in the readme, at least the output of dump...

(note that when testing with native_posix, zephyr.bin does not exist)

yonsch commented:
Thanks everyone!

attie commented on subsys/bindesc/Kconfig.build_time:
```diff
@@ -0,0 +1,106 @@
+config BINDESC_BUILD_DATE_TIME_STRING_FORMAT
+	depends on BINDESC_BUILD_DATE_TIME_STRING
+	string "Date-Time format"
+	default "%Y/%m/%d %H:%M:%S"
```
I'd have a preference for ISO 8601 timestamps and explicitly UTC... i.e: `%Y-%m-%dT%H:%M:%SZ`

Same for `BINDESC_BUILD_DATE_STRING_FORMAT` (`%Y-%m-%d`) and `BINDESC_BUILD_TIME_STRING_FORMAT` (`%H:%M:%SZ`).

attie commented on subsys/bindesc/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+
+macro(gen_build_time_int_definition def_name format)
+  if(CONFIG_BINDESC_${def_name})
+    string(TIMESTAMP ${def_name} ${format})
```
Add `UTC`?

attie commented on subsys/bindesc/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+
+macro(gen_build_time_str_definition def_name format)
+  if(CONFIG_BINDESC_${def_name})
+    string(TIMESTAMP ${def_name} ${${format}})
```
Add `UTC`?

attie reviewed (commented):
This looks really great - well done all involved, especially @yonsch.

Apologies for missing the merge - my only real comment is about timestamps... can we explicitly use UTC, and ISO 8601 format?
